crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO(william) add metrics
+// TODO(william) make configurable
+
+use either::Either;
+use futures::future::join_all;
+use futures::stream::FuturesOrdered;
+
+use sui_macros::nondeterministic;
+use sui_types::base_types::{ObjectDigest, ObjectID};
+use tracing::debug;
+use typed_store::Map;
+
+use std::path::Path;
+use std::sync::Arc;
+
+use fastcrypto::hash::MultisetHash;
+use mysten_metrics::spawn_monitored_task;
+use sui_types::accumulator::Accumulator;
+use sui_types::committee::EpochId;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio_stream::StreamExt;
+use typed_store::rocks::TypedStoreError;
+use typed_store::rocks::{DBMap, MetricConf};
+use typed_store::traits::TableSummary;
+use typed_store::traits::TypedStoreDebug;
+use typed_store_derive::DBMapUtils;
+
+use crate::authority::authority_notify_read::NotifyRead;
+
+type AccumulatorTaskBuffer = FuturesOrdered<JoinHandle<SuiResult<(Accumulator, State)>>>;
+type EndOfEpochFlag = bool;
+
+#[derive(Debug, Clone)]
+pub struct State {
+    pub effects: Vec<TransactionEffects>,
+    pub checkpoint_seq_num: CheckpointSequenceNumber,
+    pub epoch: EpochId,
+    pub end_of_epoch_flag: EndOfEpochFlag,
+}
+
+#[derive(DBMapUtils)]
+pub struct StateAccumulatorTables {
+    // TODO: implement pruning policy as most of these tables can
+    // be cleaned up at end of epoch
+
+    // Maps checkpoint sequence number to an accumulator with accumulated state
+    // only for the checkpoint that the key references. Append-only, i.e.,
+    // the accumulator is complete wrt the checkpoint
+    pub state_hash_by_checkpoint: DBMap<CheckpointSequenceNumber, Accumulator>,
+
+    // A live, append-only table representing the running root state hash, computed at
+    // arbitrary checkpoints. As such, should NEVER be relied upon for snapshot
```

Comment:
to make sure I understand - you say NEVER because we're not guaranteed to compute the accumulator at checkpoint C for all validators, right? it is the case that if all validators _did_ compute the accumulator at C, it would be the same for all of them, right?

Reply:
Correct, I'm saying don't use it because it is not guaranteed to exist for an arbitrary checkpoint. Although this comment is less relevant now that we are mapping this incremental accumulator to a checkpoint sequence number. For any sequence number `n`, state_hash_by_checkpoint[n]` should be the same for all validators, if `n` is a key at all. So it sort of enforces its own safety. I'll revise this comment